core/src/main/scala/kafka/log/LogLoader.scala

```diff
@@ -489,16 +488,16 @@ class LogLoader(
    *
    * @param segmentsToDelete The log segments to schedule for deletion
    */
-  private def removeAndDeleteSegmentsAsync(segmentsToDelete: Iterable[LogSegment]): Unit = {
-    if (segmentsToDelete.nonEmpty) {
+  private def removeAndDeleteSegmentsAsync(segmentsToDelete: java.util.Collection[LogSegment]): Unit = {
+    if (!segmentsToDelete.isEmpty) {
       // Most callers hold an iterator into the `params.segments` collection and
       // `removeAndDeleteSegmentAsync` mutates it by removing the deleted segment. Therefore,
       // we should force materialization of the iterator here, so that results of the iteration
       // remain valid and deterministic. We should also pass only the materialized view of the
       // iterator to the logic that deletes the segments.
-      val toDelete = segmentsToDelete.toList
-      info(s"Deleting segments as part of log recovery: ${toDelete.mkString(",")}")
-      toDelete.foreach { segment =>
+      val toDelete = new util.ArrayList[LogSegment](segmentsToDelete)
+      info(s"Deleting segments as part of log recovery: ${LocalLog.mkString(toDelete.iterator(), ",")}")
```

Comment:
Can you not use `Utils.join` instead of creating a new implementation here?

Reply:
Thanks for letting me know about `Utils.join`, I was not aware of it.